src/text-document.coffee

```diff
@@ -30,3 +31,31 @@ class TextDocument
       (previousLayer, transform) -> new TransformLayer(previousLayer, transform)
       @getLinesLayer()
     )
+
+  characterIndexForPosition: (position) ->
+    @sourcePositionForPosition(
+      position, @displayLayer, @bufferLayer
+    ).column
+
+  positionForCharacterIndex: (charIndex) ->
+    position = new Point(0, charIndex)
+
+    @positionForSourcePosition(position, @displayLayer, @bufferLayer)
+
+  sourcePositionForPosition: (position, targetLayer, sourceLayer) ->
```

Comment:
I am not entirely happy with naming this `targetLayer`, as it seems like we should stop there during the traversal (whereas we actually stop when we reach `sourceLayer`). Previously I used `currentLayer`, but I didn't like it too. Any ideas? 

Also, how do you feel about naming these methods exactly as we named `Layer` ones? I think it's acceptable, since from a client-side perspective they're doing the same thing but on different "elements": `TextDocument` translates positions over the whole stack of layers, while `Layer` only translates positions to and from the source layer.